### PR TITLE
Convert embed to new version

### PR DESCRIPTION
**Pull request by mhpjay422**
Why this PR?
- Convert embed to new version

### CODE REVIEW

**Comment by mhpjay422** [comment]
CleanShot 2023-06-09 at 17 32 24

File: pages/api/flatfile/create-embed-space.tsx [hunk 1]
```diff
@@ -78,14 +71,19 @@ export default async function handler(
...
-  );
+  const title = "Welcome";
+  const body = initialDocumentBody;
+  const spaceId = flatfileSpaceData?.id as string;
+
+  const addDocumentToSpaceResponse = await addDocumentToSpace({
+    title,
+    body,
+    spaceId,
+  });
```
Comment (johnmosesman) [inline pages/api/flatfile/create-embed-space.tsx:82]: Doesn't this document get created in the listener?

File: pages/api/flatfile/create-embed-space.tsx [hunk 2]
```diff
@@ -78,14 +71,19 @@ export default async function handler(
...
+
+  console.log("addDocumentToSpaceResponse", addDocumentToSpaceResponse);
 
   res.redirect(
-    `/embedded-portal?flash=success&message=Space Created!&createdSpaceId=${flatfileSpaceData.id}`
+    `/embedded-portal?flash=success&message=Space Created!&createdSpaceId=${flatfileSpaceData?.id}`
```
Comment (johnmosesman) [inline pages/api/flatfile/create-embed-space.tsx:87]: should always have an id here right?

File: pages/embedded-portal.tsx [hunk 3]
```diff
@@ -110,7 +109,7 @@ const EmbeddedPortal: NextPageWithLayout<Props> = ({
 
           res.actions.forEach((a) => {
             toast.success(a.description, {
-              id: DateTime.now().toISO(),
+              id: DateTime.now().toISO() as string,
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:112]: why's this needed?

File: pages/embedded-portal.tsx [hunk 4]
```diff
@@ -264,16 +261,57 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+  const workbookConfig = {
+    name: workbook?.name || "HCM.show Embedded Portal",
+    sheets:
+      workbook?.sheets?.map((s) => {
+        return {
+          name: s.name,
+          slug: s.config?.slug,
+          fields: s.config?.fields,
+        };
+      }) || null,
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:299]: should this be || null or just empty array?

File: lib/new-flatfile.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
+import { FlatfileClient } from "@flatfile/api";
```
Comment (johnmosesman) [inline lib/new-flatfile.ts:1]: I renamed the old flatfile file and give the new one the name `flatfile.ts`, you can use that one

Comment (mhpjay422) [reply lib/new-flatfile.ts:1]: Edited to cutover to exiting flatfile file

File: pages/api/flatfile/create-embed-space.tsx (hunk 1 above)
Comment (mhpjay422) [reply pages/api/flatfile/create-embed-space.tsx:82]: Yes. Removed

File: pages/api/flatfile/create-embed-space.tsx (hunk 2 above)
Comment (mhpjay422) [reply pages/api/flatfile/create-embed-space.tsx:87]: Yes. Edited

File: pages/embedded-portal.tsx (hunk 3 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:112]: Getting this error. Found a better way to deal with it vs using "as string"

CleanShot 2023-06-13 at 11 14 24

File: pages/embedded-portal.tsx [hunk 6]
```diff
@@ -206,17 +204,10 @@ const EmbeddedPortal: NextPageWithLayout<Props> = ({
         </div>
       </div>
 
-      {error && <div>{error}</div>}
-      {!error && showSpace && (
+      {showSpace && (
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:207]: What does the error component look like? It's customizable in the new version

File: pages/embedded-portal.tsx [hunk 7]
```diff
@@ -49,33 +52,28 @@ const EmbeddedPortal: NextPageWithLayout<Props> = ({
...
+  if (!publishableKey) {
+    console.error("Missing NEXT_PUBLIC_EMBEDDED_PUBLISHABLE_KEY env var");
+    throw "Missing NEXT_PUBLIC_EMBEDDED_PUBLISHABLE_KEY env var";
+  }
+
   const [showSpace, setShowSpace] = useState(false);
   const spaceProps = {
-    accessToken: accessToken as string,
+    publishableKey,
     environmentId: environmentToken as string,
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:65]: should check further up the chain and not do `as string` here

File: pages/embedded-portal.tsx [hunk 8]
```diff
@@ -264,16 +254,61 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+  const spaceData = await getSpace({
+    workflow: WorkflowType.Embedded,
+    spaceId: (existingSpace?.flatfileData as unknown as FlatfileSpaceData).id,
+  });
+
+  console.log("spaceData", spaceData);
+
+  const workbook = await getWorkbook({
+    workflow: WorkflowType.Embedded,
+    workbookId: spaceData?.primaryWorkbookId as string,
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:282]: is primaryWorkbookId not guaranteed here?

File: pages/embedded-portal.tsx (hunk 7 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:65]: removed

File: pages/embedded-portal.tsx [hunk 9]
```diff
@@ -253,7 +244,6 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
 
   // console.log("existingSpace", existingSpace);
 
-  const accessToken = await getAccessToken();
   const environmentToken = process.env.EMBEDDED_ENVIRONMENT_ID;
```
[the remark this answers is not in the document]
Comment (mhpjay422) [reply pages/embedded-portal.tsx:247]: Its not yelling anymore so going to skip this conditional

File: pages/embedded-portal.tsx (hunk 8 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:282]: primaryWorkbookId is not typed onto the space. On the config, the space is getting updated with the primaryWorkbookId
CleanShot 2023-06-13 at 17 37 07
CleanShot 2023-06-13 at 17 37 18@2x

File: pages/embedded-portal.tsx [hunk 10]
```diff
@@ -264,16 +254,61 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+  });
+
+  console.log("spaceData", spaceData);
+
+  const workbook = await getWorkbook({
+    workflow: WorkflowType.Embedded,
+    workbookId: spaceData?.primaryWorkbookId as string,
+  });
+
+  console.log("workbook", JSON.stringify(workbook, null, 2));
```
[the remark this answers is not in the document]
Comment (mhpjay422) [reply pages/embedded-portal.tsx:285]: added

**Comment by mhpjay422** [comment]
CleanShot 2023-06-14 at 11 16 51

File: pages/embedded-portal.tsx [hunk 11]
```diff
@@ -26,20 +26,33 @@ import StepList, { Step } from "../components/shared/step-list";
...
 import { useFlashMessages, useOnClickOutside } from "../lib/hooks/usehooks";
+import { Flatfile } from "@flatfile/api";
+import { WorkflowType, getSpace, getWorkbook } from "../lib/flatfile";
+
+type WorkbookConfig = {
+  name: string;
+  labels?: string[];
+  spaceId: Flatfile.SpaceId;
+  environmentId: Flatfile.EnvironmentId;
+  /** Sheets must be <= 50 */
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:37]: why < 50?

File: pages/embedded-portal.tsx [hunk 12]
```diff
@@ -26,20 +26,33 @@ import StepList, { Step } from "../components/shared/step-list";
...
+  sheets?: Flatfile.SheetConfig[];
+  actions?: Flatfile.Action[];
+};
 
 interface Props {
   accessToken: string;
   environmentToken: string;
   lastSyncedAt?: string;
   existingSpace: Space;
+  workbookConfig: Pick<WorkbookConfig, "name" | "sheets" | "actions">;
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:47]: I think there's a type from the package for this, it was like CreateWorkbookConfig or something?

File: pages/embedded-portal.tsx [hunk 13]
```diff
@@ -26,20 +26,33 @@ import StepList, { Step } from "../components/shared/step-list";
 import Workspace from "../components/embedded-portal/workspace";
 import { theme } from "../lib/theme";
 import { useFlashMessages, useOnClickOutside } from "../lib/hooks/usehooks";
+import { Flatfile } from "@flatfile/api";
+import { WorkflowType, getSpace, getWorkbook } from "../lib/flatfile";
+
+type WorkbookConfig = {
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:32]: Do we need to specify this type or can we use one from the package?

File: pages/embedded-portal.tsx [hunk 14]
```diff
@@ -264,16 +274,83 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+
+  let workbook = await getWorkbook({
+    workflow: WorkflowType.Embedded,
+    workbookId: spaceData?.primaryWorkbookId!,
+  });
+
+  console.log("workbook", workbook);
+
+  const timeInFive = new Date();
+  timeInFive.setSeconds(timeInFive.getSeconds() + 5);
```
[the remark this answers is not in the document]
Comment (johnmosesman) [reply pages/embedded-portal.tsx:308]: Same for below, there's probably a comparison operator

File: pages/embedded-portal.tsx [hunk 15]
```diff
@@ -264,16 +274,83 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+
+    workbook = await getWorkbook({
+      workflow: WorkflowType.Embedded,
+      workbookId: spaceData?.primaryWorkbookId!,
+    });
+  }
+
+  if (!workbook) {
+    console.log("Unable to get workbook");
+    throw "Unable to get workbook";
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:324]: What happens if this throws? Should we redirect with an error message instead?

File: pages/embedded-portal.tsx [hunk 16]
```diff
@@ -264,16 +274,83 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+    name: workbook?.name || "HCM.show Embedded Portal",
+    sheets:
+      workbook?.sheets?.map((s) => {
+        return {
+          name: s.name,
+          slug: s.config?.slug,
+          fields: s.config?.fields,
+        };
+      }) || [],
+    actions: workbook?.actions || [],
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:337]: Do we need all these `?` and `[]`?

File: pages/embedded-portal.tsx (hunk 11 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:37]: Pasted from the docs. Removing the comment

File: pages/embedded-portal.tsx (hunk 16 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:337]: Cleaned it up. Sheets on workbook is typed as a optional prop. 
CleanShot 2023-06-14 at 13 54 43

File: pages/embedded-portal.tsx (hunk 13 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:32]: It should be there, but I cant import it from Flatfile. Its there in the docs imported from Flatfile. Weird.

Comment (mhpjay422) [reply pages/embedded-portal.tsx:32]: CleanShot 2023-06-14 at 14 00 53

File: pages/embedded-portal.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (mhpjay422) [reply pages/embedded-portal.tsx:308]: Switched to luxon

File: pages/embedded-portal.tsx (hunk 13 above)
Comment (johnmosesman) [reply pages/embedded-portal.tsx:32]: try `Flatfile.CreateWorkbookConfig` ?

File: pages/embedded-portal.tsx [hunk 17]
```diff
@@ -49,33 +61,32 @@ const EmbeddedPortal: NextPageWithLayout<Props> = ({
...
+
+  if (!publishableKey) {
+    console.error("Missing NEXT_PUBLIC_EMBEDDED_PUBLISHABLE_KEY env var");
+    throw "Missing NEXT_PUBLIC_EMBEDDED_PUBLISHABLE_KEY env var";
+  }
+
   const [showSpace, setShowSpace] = useState(false);
+  const error = (error: string) => (
+    <div className="text-white text-lg font-semibold">{error}</div>
+  );
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:74]: i'd give this a background with an error object or something and give a standard error, "An error occurred opening the portal."

Comment (johnmosesman) [reply pages/embedded-portal.tsx:74]: + console.error with the actual error

File: pages/embedded-portal.tsx [hunk 18]
```diff
@@ -264,16 +273,87 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+      workbookId: spaceData?.primaryWorkbookId!,
+    });
+  }
+
+  if (!workbook) {
+    console.log("Unable to get workbook");
+    return {
+      redirect: {
+        // TODO: add a flash message on the home page
+        destination: "/home",
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:325]: back to /embedded?

File: pages/embedded-portal.tsx [hunk 19]
```diff
@@ -264,16 +273,87 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
+  let workbook = await getWorkbook({
+    workflow: WorkflowType.Embedded,
+    workbookId: spaceData?.primaryWorkbookId!,
+  });
+
+  console.log("workbook", workbook);
+
+  const timeInFive = DateTime.now().plus({ seconds: 5 });
+
+  while (!(DateTime.now().diff(timeInFive).toMillis() > 0 || workbook)) {
```
Comment (johnmosesman) [inline pages/embedded-portal.tsx:308]: DateTime.now() > timeInFive  ?

https://moment.github.io/luxon/#/math?id=comparing-datetimes

File: pages/embedded-portal.tsx (hunk 13 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:32]: edited to new

File: pages/embedded-portal.tsx (hunk 15 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx]: edited to redirect

File: pages/embedded-portal.tsx (hunk 19 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:308]: edited

File: pages/embedded-portal.tsx (hunk 17 above)
Comment (mhpjay422) [reply pages/embedded-portal.tsx:74]: edited

Comment (mhpjay422) [reply pages/embedded-portal.tsx:74]: CleanShot 2023-06-14 at 19 06 38

Comment (mhpjay422) [reply pages/embedded-portal.tsx:74]: CleanShot 2023-06-14 at 19 08 32

**Pull request merged**